### PR TITLE
feat(payment_charges): add support for collecting and refunding charges on payments

### PR DESCRIPTION
**Pull request by kashif-m**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Detailed description - https://github.com/juspay/hyperswitch/issues/4659

This PR adds functionality to create charges on payment intents using Stripe
- create `charges` on payment_intent
- store `charge_id` in payment_attempt
- use `charge_id` for refunds
   - Direct
   ```
    "charges": {
        "charge_id": "ch_3PJDz8Ihl7EEkW0O16Wetnl5",
        "revert_platform_fee": true
    }
    ```
   - Destination
   ```
    "charges": {
        "charge_id": "ch_3PJDz8Ihl7EEkW0O16Wetnl5",
        "revert_platform_fee": true,
        "revert_transfer": true
    }
_Note_ - charges on mandates is not added in this PR.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

<!--
Provide links to the files with corresponding changes.

Following are the paths where you can find config files:
1. `config`
2. `crates/router/src/configs`
3. `loadtest/config`
-->


## Motivation and Context
<!--
Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.

If you don't have an issue, we'd recommend starting with one first so the PR
can focus on the implementation (unless it is an obvious bug or documentation fix
that will have little conversation).
-->


## How did you test it?
<!--
Did you write an integration/unit/API test to verify the code changes?
Or did you test this change manually (provide relevant screenshots)?
-->
Tested locally using postman collection.

**Steps for testing**
- Make sure Stripe Connect is enabled on your Stripe account
- Onboard a Standard user by initiating an account link from Stripe's dashboard
- Make sure merchant account is created and Stripe is added as a payment connector
- Create a direct charge on payment by adding below field in `/payment_intents` request
```
"charges": {
        "charge_type": "direct",
        "fees": 123,
        "transfer_account_id": "acct_1PDftAIhl7EEkW0O"
}
```
- Create a destination charge on payment by adding below field in `/payment_intents` request
```
"charges": {
        "charge_type": "destination",
        "fees": 123,
        "transfer_account_id": "acct_1PDftAIhl7EEkW0O"
}
```

### Direct charges
- ref - https://docs.stripe.com/connect/direct-charges
- these help in directly creating charges when customer is paying to the connected account
- charges are created, money is split (Stripe fees + reseller’s fees) and the remaining amount is transferred to connected account’s balance

### Destination charges
- ref - https://docs.stripe.com/connect/destination-charges
- charges are created on the reseller’s account
- reseller decides whether some of all of those funds are to be transferred to the connected account
- reseller’s account is used for debiting the cost of Stripe fees, refunds and chargebacks
- only supported when both reseller and connected accounts belong to the same country



## Checklist
<!-- Put an `x` in the boxes that apply -->

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/api_models/src/enums.rs [hunk 1]
```diff
@@ -590,3 +590,47 @@ pub fn convert_pm_auth_connector(connector_name: &str) -> Option<PmAuthConnector
...
+    Eq,
+    PartialEq,
+    serde::Deserialize,
+    serde::Serialize,
+    strum::Display,
+    strum::EnumString,
+    ToSchema,
+    Hash,
+)]
+pub enum PaymentChargeType {
```
[the remark this answers is not in the document]
Comment (kashif-m) [reply crates/api_models/src/enums.rs:606]: It's not declared twice, I could not find this in payments.rs :thinking: 

Can you send a link to this declaration in payments.rs?

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -467,6 +467,29 @@ pub struct PaymentsRequest {
...
+    /// Fee information to be charged on the payment being collected
+    pub charges: Option<PaymentChargeRequest>,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone, ToSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum PaymentChargeRequest {
+    #[serde(untagged)]
+    Stripe(StripeChargeRequest),
+}
```
[the remark this answers is not in the document]
Comment (kashif-m) [reply crates/api_models/src/payments.rs]: Made it a generic request

File: crates/router/src/connector/stripe/transformers.rs [hunk 3]
```diff
@@ -157,6 +157,27 @@ pub struct PaymentIntentRequest {
...
+#[derive(Debug, Eq, PartialEq, Serialize)]
+pub struct DirectCharges {
+    pub application_fee_amount: i64,
+}
+
+#[derive(Debug, Eq, PartialEq, Serialize)]
+pub struct DestinationCharges {
+    pub application_fee_amount: i64,
+    #[serde(rename = "transfer_data[destination]")]
+    pub destination_account_id: String,
```
Comment (SamraatBansal) [inline crates/router/src/connector/stripe/transformers.rs:180]: Put the fields `application_fee_amount` and `destination_account_id`  in the `PaymentIntentRequest`

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

File: crates/cards/dep.dot
[the remark this answers is not in the document]
Comment (kashif-m) [reply crates/cards/dep.dot:1]: No, my bad

**Review by SamraatBansal — APPROVED** [review]
LGTM! Approving on behalf of connectors

**Review by SamraatBansal — DISMISSED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Review by ThisIsMani — DISMISSED** [review]
Dashboard specific changes looks fine.

**Review by SamraatBansal — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Pull request merged**